### PR TITLE
feat: Feature to have widget features

### LINKED ISSUE #17976: [Task]: Create platform structure for adding widget features
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

Widgets can be enhanced by the platform using simple configs. To achieve this, widgets need to enhance the following
- [ ] Property pane configuration
- [ ] Default properties

The following are the subtasks to enable this
- [ ] Create a Widget features registry
- [ ] Enhance property pane configurations based on feature registry and widget feature configs
- [ ] Enhance default properties for widgets based on feature registry and widget feature configs.

To validate this change we'll be using the auto height feature
- [ ] Add auto height to widget features registry
- [ ] Add auto height default properties to widget features map
- [ ] Add auto height property pane enhancements to widget features map
- [ ] Add widget helper utilities for auto height

### PR DESCRIPTION
**Pull request by riodeuno**
## Description

- [ ] Adds structure for adding widget features
- [ ] Currently added details for the auto height feature

Fixes #17976 

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Jest PENDING

### Test Plan
NA

### Issues raised during DP testing
NA

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/utils/WidgetFeatures.ts [hunk 1]
```diff
@@ -1,13 +1,24 @@
...
+import { WidgetHeightLimits } from "constants/WidgetConstants";
+import { ValidationTypes } from "constants/WidgetValidation";
 import { WidgetProps } from "widgets/BaseWidget";
+import { WidgetConfiguration } from "widgets/constants";
 
-export interface WidgetFeatures {
-  dynamicHeight: boolean;
+import { AutocompleteDataType } from "./autocomplete/TernServer";
+
+export enum RegisteredWidgetFeatures {
```
Comment (riodeuno) [inline app/client/src/utils/WidgetFeatures.ts:12]: Split this file (unless this gives us a cyclic dependency errors) and add unit tests.

File: app/client/src/constants/WidgetConstants.tsx [hunk 2]
```diff
@@ -152,3 +152,8 @@ export const WIDGET_DSL_STRUCTURE_PROPS = {
 export type TextSize = keyof typeof TextSizes;
 
 export const DEFAULT_FONT_SIZE = THEMEING_TEXT_SIZES.base;
+
+export const WidgetHeightLimits = {
+  MAX_HEIGHT_IN_ROWS: 9000,
```
Comment (marks0351) [inline app/client/src/constants/WidgetConstants.tsx:157]: can we add comments why 9000 and 4?

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/src/utils/WidgetFeatures.ts [hunk 3]
```diff
@@ -18,68 +29,326 @@ export enum DynamicHeight {
...
    This function checks if the `dynamicHeight` property is enabled
    and returns true if disabled, and false if enabled.
 */
 export function hideDynamicHeightPropertyControl(props: WidgetProps) {
-  return props.dynamicHeight !== DynamicHeight.HUG_CONTENTS;
+  return props.dynamicHeight !== DynamicHeight.AUTO_HEIGHT_WITH_LIMITS;
+}
+
+function validateMinHeight(value: unknown, props: WidgetProps) {
+  const _value: number = parseInt(value as string, 10);
```
[the remark this answers is not in the document]
Comment (riodeuno) [reply app/client/src/utils/WidgetFeatures.ts:126]: No, this is the [parseInt's second parameter (base or radix)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt). We can also use hex or binary, etc, as well.

File: app/client/src/utils/WidgetFeatures.ts [hunk 4]
```diff
@@ -18,68 +29,326 @@ export enum DynamicHeight {
...
+  const _maxHeight: number = parseInt(props.maxDynamicHeight as string, 10);
+
+  if (isNaN(_value) || _value < 4) {
```
[the remark this answers is not in the document]
Comment (riodeuno) [reply app/client/src/utils/WidgetFeatures.ts:129]: This one should be `WidgetHeightLimits.MIN_HEIGHT`. I'll fix this.

File: app/client/src/utils/WidgetFeatures.ts (hunk 1 above)
Comment (riodeuno) [reply app/client/src/utils/WidgetFeatures.ts:12]: Adding unit tests, but we won't split this in this PR.

Comment (riodeuno) [reply app/client/src/utils/WidgetFeatures.ts:12]: Task created: https://github.com/appsmithorg/appsmith/issues/17965

**Review by marks0351 — DISMISSED** [review]

**Review by marks0351 — APPROVED** [review]

**Pull request merged**